ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,482 @@
+                match known_bits.borrow_mut().feed(opt, inst) {
+                    OptOutcome::Rewritten(new_inst) => {
+                        strength_fold.borrow_mut().feed(opt, new_inst)
+                    }
+                    rest => rest,
+                }
+            },
+            |opt, iidx, inst| known_bits.borrow_mut().inst_committed(opt, iidx, inst),
+            ptn,
+        );
+    }
+
+    #[test]
+    fn opt_and() {
+        // any = any & 01
+        // any = any & 11 <-- should be removed
+        test_known_bits(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = 1
+          %2: i8 = 3
+          %3: i8 = and %0, %1
+          %4: i8 = and %3, %2
+          blackbox %4
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = 1
+          %2: i8 = 3
+          %3: i8 = and %0, %1
+          blackbox %3
+        ",
+        );
+
+        // any = any & 11
+        // any = any & 01 <-- should not be removed
+        test_known_bits(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = 3
+          %2: i8 = 1
+          %3: i8 = and %0, %1
+          %4: i8 = and %3, %2
+          blackbox %4
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = 3
+          %2: i8 = 1
+          %3: i8 = and %0, %1
+          %4: i8 = and %3, %2
+          blackbox %4
+        ",
+        );
+
+        // any & any
+        test_known_bits(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = arg [reg]
+          %2: i8 = and %0, %1
+          blackbox %2
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = arg
+          %2: i8 = and %0, %1
+          blackbox %2
+        ",
+        );
+    }
+
+    #[test]
+    fn opt_or() {
+        // Test bitwise or
+        // any = any | 01
+        // any = any | 10
+        // any = any | 11 <-- should be removed
+        test_known_bits(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = 1
+          %2: i8 = 2
+          %3: i8 = 3
+          %4: i8 = or %0, %1
+          %5: i8 = or %4, %2
+          %6: i8 = or %5, %3
+          blackbox %6
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = 1
+          %2: i8 = 2
+          %3: i8 = 3
+          %4: i8 = or %0, %1
+          %5: i8 = or %4, %2
+          blackbox %5
+        ",
+        );
+
+        // Test bitwise or
+        // any = any | 01
+        // any = any | 10  <-- should not be removed
+        test_known_bits(
+            "
+          %0: i8 = arg [reg]
+          %1: i8 = 1
+          %2: i8 = 2
+          %3: i8 = 3
+          %4: i8 = or %0, %1
+          %5: i8 = or %4, %2
+          blackbox %5
+        ",
+            "
+          %0: i8 = arg
+          %1: i8 = 1
+          %2: i8 = 2
+          %3: i8 = 3
+          %4: i8 = or %0, %1
+          %5: i8 = or %4, %2
+          blackbox %5
+        ",
+        );
+    }
+
+    #[test]
+    fn opt_and_or() {
```

Comment:
I'm not convinced by this test because scaling it's going to be _really_ hard as we analyse more IR instructions! I'd probably remove it.